addons/payment/tests/test_payment_token.py

```diff
@@ -28,9 +28,21 @@ def test_cannot_assign_token_to_public_partner(self):
         with self.assertRaises(ValidationError):
             token.partner_id = self.public_user.partner_id
 
-    def test_token_cannot_be_unarchived(self):
-        """ Test that unarchiving disabled tokens is forbidden. """
+    def test_unarchiving_token_requires_active_provider(self):
+        """ Test that unarchiving disabled tokens is forbidden if the provider is disabled. """
         token = self._create_token(active=False)
+        token.provider_id.state = 'disabled'
+        with self.assertRaises(UserError):
+            token.active = True
+
+    def test_unarchiving_token_requires_active_payment_method(self):
+        """ Test that unarchiving disabled tokens is forbidden if the method is disabled. """
+        token = self._create_token(active=False)
+        token.payment_method_id.active = False
+        token.provider_id.state = 'enabled'
+        with self.assertRaises(UserError):
+            token.active = True
+        token.provider_id.state = 'disabled'
```

Comment:
No need to test this behavior based on the provider state as it's already done with `test_unarchiving_token_requires_active_provider`.
```suggestion
```